proposals/lip-0055.md

```diff
@@ -105,16 +129,45 @@ The full processing of a block is organized as follows.
+
+1. **Header initialization**: Block header properties that require access to the state store *before* any state transitions implied by the block are executed are inserted in this stage.
+2. **Assets initialization**: Each module can insert information in the block assets *before* any state transitions implied by the block are executed.
+3. **Before transactions execution**: Each module can define protocol logic that is executed *before* the transactions contained in the block are processed. 
+After this stage has been completed, transactions are selected one-by-one from a transaction pool. The transaction processing stages (stages 4 to 8) are repeated for each transaction selected. If stages 4, 5, 6, and 8 are executed succesfully, the transaction is valid and it is included in the block, otherwise it is invalid and therefore discarded.
+4. **Transaction verification**: Each module can define protocol logic that verifies a transaction, possibly by accessing the state store. If an error occurs, the transaction is invalid and it is not included in the block.
+5. **Command verification**: The command corresponding to the `moduleID`-`commandID` combination is verified. If an error occurs, the transaction is invalid and it is not included in the block.
+6. **Before command execution**: Each module can define protocol logic that is processed *before* the command has been executed. If an error occurs, the transaction is invalid, it is not included in the block, all state transitions induced by the transaction are reverted. In that case, the block generation continues with stage 4 for another transaction from the transaction pool or stage 9.
+7. **Command execution**: The command corresponding to the `moduleID`-`commandID` combination is executed. If an error occurs, the transaction is failed and all state transitions performed in this stage are reverted. In any case, afterwards the processing continues with the next stage.
+8. **After command execution**: Each module can define protocol logic that is processed *after* the command has been executed. If an error occurs, the transaction is invalid, it is not included in the block and all state transitions induced by the transaction performed up to this stage are reverted. In that case, the block generation continues with stage 4 for another transaction from the transaction pool or stage 9.
+9. **After transactions execution**: Each module can define protocol logic that is executed *after* all the transactions contained in the block have been processed.
+10. **Assets finalization**: Each module can update or insert new information in the block assets *after* the state transitions implied by the block are executed.
+11. **Header finalization**: Block header properties, which require accessing the state store *after* all state transitions implied by the block have been executed, are inserted.
+12. **Block processing**: The block goes through the [block processing stages](#block-processing-stages).
```

Comment:
I would suggest to no put this here as a stage, but mention as a commented below this list that a generated block would afterwards be passed to the block processing and will go through the stages defined above. This would yield a cleaner separation in my opinion:

- block generation: produce a block that induces a valid state transition and will be accepted by any node in the network
- block processing: executing the state changes implied by a block

Otherwise, block generation would always be block processing at the same time.

Reply:
The specifications above are also currently incorrect, independently of that. Because as written now, after the header finalization the state changes implied by the transactions are already performed. So maybe step 12 should be reverting the state back to before the block processing?

Reply:
Looking at the stages again, I am wondering whether the asset finalization actually makes sense together with the block generation process above because all state changes are already applied before.

Specifically, consider the following example: Assume the block reward is in the asset and it would be reduced only in the asset finalization stage. Then the state root computed in the block generation would be different from the one in the block processing (as in the block generation 1 LSK would be credited to the block generator but only 0.25 LSK in the block processing).

Reply:
As concluded, we remove the asset finalization. This comment can therefore be resolved.